### PR TITLE
[AArch64] Replace LLVM IR function attributes for PSTATE.ZA.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
Since https://github.com/ARM-software/acle/pull/276 the ACLE
defines attributes to better describe the use of a given SME state.

Previously the attributes merely described the possibility of it being
'shared' or 'preserved', whereas the new attributes have more semantics
and also describe how the data flows through the program.

For ZT0 we already had to add new LLVM IR attributes:
* aarch64_new_zt0
* aarch64_in_zt0
* aarch64_out_zt0
* aarch64_inout_zt0
* aarch64_preserves_zt0

We have now done the same for ZA, such that we add:
* aarch64_new_za       (previously `aarch64_pstate_za_new`)
* aarch64_in_za        (more specific variation of `aarch64_pstate_za_shared`)
* aarch64_out_za       (more specific variation of `aarch64_pstate_za_shared`)
* aarch64_inout_za     (more specific variation of `aarch64_pstate_za_shared`)
* aarch64_preserves_za (previously `aarch64_pstate_za_shared, aarch64_pstate_za_preserved`)

This explicitly removes 'pstate' from the name, because with SME2 and
the new ACLE attributes there is a difference between "sharing ZA" (sharing
the ZA matrix register with the caller) and "sharing PSTATE.ZA" (sharing
either the ZA or ZT0 register, both part of PSTATE.ZA with the caller).

### CODE REVIEW

File: llvm/docs/AArch64SME.rst [hunk 1]
```diff
@@ -22,26 +22,32 @@ Below we describe the LLVM IR attributes and their relation to the C/C++
...
+    is used for functions with ``__arm_streaming_compatible``
 
 ``aarch64_pstate_sm_body``
-  is used for functions with ``__attribute__((arm_locally_streaming))`` and is
+  is used for functions with ``__arm_locally_streaming`` and is
   only valid on function definitions (not declarations)
 
-``aarch64_pstate_za_new``
-  is used for functions with ``__attribute__((arm_new_za))``
+``aarch64_pstate_new_za``
```
Comment (c-rhodes) [inline llvm/docs/AArch64SME.rst:34]:
```suggestion
``aarch64_new_za``
```

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 2]
```diff
@@ -1098,11 +1098,15 @@ LogicalResult ModuleTranslation::convertOneFunction(LLVMFuncOp func) {
...
+    llvmFunc->addFnAttr("aarch64_new_za");
+  else if (func.getArmInZa())
+    llvmFunc->addFnAttr("aarch64_in_za");
+  else if (func.getArmOutZa())
+    llvmFunc->addFnAttr("aarch64_out_za");
+  else if (func.getArmInoutZa())
+    llvmFunc->addFnAttr("aarch64_inout_za");
   if (func.getArmPreservesZa())
-    llvmFunc->addFnAttr("aarch64_pstate_za_preserved");
+    llvmFunc->addFnAttr("aarch64_preserves_za");
```
Comment (c-rhodes) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1109]: this should be `else if`, I missed this from https://llvm.org/docs/AArch64SME.html#restrictions-on-attributes
> It is not allowed for a function to be decorated with both aarch64_pstate_za_new and aarch64_pstate_za_preserved.

when I originally reviewed this. I see you've fixed it for import, please could you fix it here as well?

**Review by c-rhodes — COMMENTED** [review]
MLIR changes LGTM cheers, just left one comment for something I missed when the preserved attr was added recently.

Also, in the commit message the LLVM IR attributes are prefixed with `arm`, should be `aarch64`.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 2 above)
Comment (MacDue) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1109]: (it used to be that `aarch64_pstate_za_preserved` was used with `aarch64_pstate_za_shared`, which is why the import was originally like this)

**Review by c-rhodes — APPROVED** [review]

**Pull request merged**